src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -89,10 +90,10 @@ public void onEnter() {
                 .collect(Collectors.toMap(et -> et, et -> new StreamKpiData()));
 
         idleStreamWatcher = new IdleStreamWatcher(getParameters().commitTimeoutMillis * 2);
-
         this.eventConsumer = getContext().getTimelineService().createEventConsumer(null);
 
         recreateTopologySubscription();
```

Comment:
The exception is thrown later (after sending the headers), as the switch to the cleanup state is done by adding the task to the `taskQueue`. Hence, it executes after all other scheduled tasks are completed. For fixing this (handling switch to cleanup state as quickly as the exception occurs, I have two suggestions

1. We switch to cleanup state as soon as the exception occurs and don't schedule it using the task queue.
2. We schedule it using the task queue, but use a priority queue to set it's priority as maximum

Reply:
I added the solution 1 as part of commit `4217a3d4`.